lib/Parse/ParseDecl.cpp

```diff
@@ -4909,6 +4910,104 @@ ParserStatus Parser::parseTypeAttribute(TypeAttributes &Attributes,
   return makeParserSuccess();
 }
 
+static llvm::Optional<LifetimeDependenceKind>
+getLifetimeDependenceKind(const Token &T) {
+  if (T.isContextualKeyword("_copy")) {
+    return LifetimeDependenceKind::Copy;
+  }
+  if (T.isContextualKeyword("_consume")) {
+    return LifetimeDependenceKind::Consume;
+  }
+  if (T.isContextualKeyword("_borrow")) {
+    return LifetimeDependenceKind::Borrow;
+  }
+  if (T.isContextualKeyword("_mutate")) {
+    return LifetimeDependenceKind::Mutate;
+  }
+  return llvm::None;
+}
+
+ParserStatus Parser::parseLifetimeDependenceSpecifiers(
+    SmallVectorImpl<LifetimeDependenceSpecifier> &specifierList) {
+  ParserStatus status;
+  // TODO: Add fixits for diagnostics in this function.
+  do {
+    auto lifetimeDependenceKind = getLifetimeDependenceKind(Tok);
+    if (!lifetimeDependenceKind.has_value()) {
+      break;
+    }
+    // consume the lifetime dependence kind
+    consumeToken();
+
+    if (!Tok.isFollowingLParen()) {
+      diagnose(Tok, diag::expected_lparen_after_lifetime_dependence);
+      status.setIsParseError();
+      continue;
+    }
+    // consume the l_paren
+    auto lParenLoc = consumeToken();
+    SourceLoc rParenLoc;
+    bool foundParamId = false;
```

Comment:
Do you actually need to keep track of `foundParamId`? Couldn’t you just check `specifierList.size() == 0`?

Reply:
`_copy(<some random unexpected thing>)` should not emit 2 diagnostics. So I think `specifierList.empty()` is not enough.